File: scripts/refactor_db_names.js
const fs = require('fs');
const path = require('path');

const replacements = {
  'persons': 'personas',
  'person_notes': 'notas_persona',
  'rescue_zones': 'zonas_rescate',
  'help_requests': 'solicitudes_ayuda',
  'volunteers': 'voluntarios',
  'volunteer_checkins': 'asistencias_voluntario',
  'case_signals': 'señales_caso',
  'moderation_log': 'bitacora_moderacion',
  'pet_reports': 'reportes_mascotas',
  'shelter_reports': 'reportes_refugios',
  'hospital_lists': 'listas_hospital',
  'hospital_list_items': 'elementos_lista_hospital',

  // Database Columns
  'created_at': 'creado_en',
  'updated_at': 'actualizado_en',
  'expires_at': 'expira_en',
  'author_name': 'nombre_reportante',
  'author_contact': 'contacto_reportante',
  'author_relation': 'relacion_reportante',
  'cedula_identidad': 'cedula_identidad',
  'full_name': 'nombre_completo',
  'alternate_names': 'nombres_alternativos',
  'sex': 'sexo',
  'age_estimated': 'edad_estimada',
  'date_of_birth': 'fecha_nacimiento',
  'physical_desc': 'descripcion_fisica',
  'clothing_desc': 'descripcion_vestimenta',
  'photo_url': 'url_foto',
  'last_seen_address': 'ultima_direccion_conocida',
  'location_zone': 'zona_ubicacion',
  'location_normalized': 'ubicacion_normalizada',
  'last_seen_lat': 'latitud_visto_ultima_vez',
  'last_seen_lng': 'longitud_visto_ultima_vez',
  'last_seen_at': 'fecha_visto_ultima_vez',
  'last_seen_context': 'contexto_visto_ultima_vez',
  'found_at': 'encontrado_en',
  'found_address': 'direccion_encontrado',
  'found_lat': 'latitud_encontrado',
  'found_lng': 'longitud_encontrado',
  'found_notes': 'notas_hallazgo',
  'verified': 'verificado',
  'verified_by': 'verificado_por',
  'verified_at': 'fecha_verificacion',
  'spam': 'spam',
  'duplicate_of': 'duplicado_de',
  'potential_duplicate_of': 'posible_duplicado_de',
  'is_deleted': 'esta_eliminado',
  'deleted_by': 'eliminado_por',
  'deleted_at': 'fecha_eliminacion',
  'deletion_reason': 'motivo_eliminacion',
  // status and source are too common, we will handle them with extra care later or skip if too risky
};

function walk(dir) {
  let results = [];
  const list = fs.readdirSync(dir);
  list.forEach(file => {
    file = path.join(dir, file);
    const stat = fs.statSync(file);
    if (stat && stat.isDirectory()) {
      results = results.concat(walk(file));
    } else {
      if (file.endsWith('.ts') || file.endsWith('.tsx')) {
        results.push(file);
      }
    }
  });
  return results;
}

const files = walk('./src').concat(walk('./scripts')).filter(f => !f.includes('refactor_db_names.js'));

for (const file of files) {
  let content = fs.readFileSync(file, 'utf8');
  let changed = false;
  for (const [oldName, newName] of Object.entries(replacements)) {
    const regex = new RegExp(`\\b${oldName}\\b`, 'g');
    if (regex.test(content)) {
      content = content.replace(regex, newName);
      changed = true;
    }
  }
  
  // Custom manual replacements for status and source only in SQL and specific object contexts
  // e.g. status -> estado_actual
  content = content.replace(/persons\.status/g, 'personas.estado_actual');
  content = content.replace(/p\.status/g, 'p.estado_actual');
  content = content.replace(/SET status/g, 'SET estado_actual');
  content = content.replace(/status =/g, 'estado_actual =');
  content = content.replace(/status IN/g, 'estado_actual IN');
  content = content.replace(/status NOT IN/g, 'estado_actual NOT IN');
  content = content.replace(/row\.status/g, 'row.estado_actual');
  content = content.replace(/status:/g, 'estado_actual:'); // risky, we'll fix { status: 200 } back to normal below
  content = content.replace(/estado_actual: 200/g, 'status: 200');
  content = content.replace(/estado_actual: 400/g, 'status: 400');
  content = content.replace(/estado_actual: 401/g, 'status: 401');
  content = content.replace(/estado_actual: 403/g, 'status: 403');
  content = content.replace(/estado_actual: 404/g, 'status: 404');
  content = content.replace(/estado_actual: 500/g, 'status: 500');
  
  // Also res.status(
  content = content.replace(/res\.estado_actual/g, 'res.status');
  content = content.replace(/response\.estado_actual/g, 'response.status');
  
  if (content !== fs.readFileSync(file, 'utf8')) {
    fs.writeFileSync(file, content, 'utf8');
    console.log(`Refactored ${file}`);
  }
}

File: scripts/refactor_db_names.py
import os
import re

replacements = {
    r'\bpersons\b': 'personas',
    r'\bperson_notes\b': 'notas_persona',
    r'\brescue_zones\b': 'zonas_rescate',
    r'\bhelp_requests\b': 'solicitudes_ayuda',
    r'\bvolunteers\b': 'voluntarios',
    r'\bvolunteer_checkins\b': 'asistencias_voluntario',
    r'\bcase_signals\b': 'señales_caso',
    r'\bmoderation_log\b': 'bitacora_moderacion',
    r'\bpet_reports\b': 'reportes_mascotas',
    r'\bshelter_reports\b': 'reportes_refugios',
    r'\bhospital_lists\b': 'listas_hospital',
    r'\bhospital_list_items\b': 'elementos_lista_hospital',

    r'\bsource\b': 'origen',
    r'\bsource_url\b': 'url_origen',
    r'\bcreated_at\b': 'creado_en',
    r'\bupdated_at\b': 'actualizado_en',
    r'\bexpires_at\b': 'expira_en',
    r'\bauthor_name\b': 'nombre_reportante',
    r'\bauthor_contact\b': 'contacto_reportante',
    r'\bauthor_relation\b': 'relacion_reportante',
    r'\bcedula_identidad\b': 'cedula_identidad',
    r'\bfull_name\b': 'nombre_completo',
    r'\balternate_names\b': 'nombres_alternativos',
    r'\bsex\b': 'sexo',
    r'\bage_estimated\b': 'edad_estimada',
    r'\bdate_of_birth\b': 'fecha_nacimiento',
    r'\bphysical_desc\b': 'descripcion_fisica',
    r'\bclothing_desc\b': 'descripcion_vestimenta',
    r'\bphoto_url\b': 'url_foto',
    r'\blast_seen_address\b': 'ultima_direccion_conocida',
    r'\blocation_zone\b': 'zona_ubicacion',
    r'\blocation_normalized\b': 'ubicacion_normalizada',
    r'\blast_seen_lat\b': 'latitud_visto_ultima_vez',
    r'\blast_seen_lng\b': 'longitud_visto_ultima_vez',
    r'\blast_seen_at\b': 'fecha_visto_ultima_vez',
    r'\blast_seen_context\b': 'contexto_visto_ultima_vez',
    r'\bfound_at\b': 'encontrado_en',
    r'\bfound_address\b': 'direccion_encontrado',
    r'\bfound_lat\b': 'latitud_encontrado',
    r'\bfound_lng\b': 'longitud_encontrado',
    r'\bfound_notes\b': 'notas_hallazgo',
    r'\bverified\b': 'verificado',
    r'\bverified_by\b': 'verificado_por',
    r'\bverified_at\b': 'fecha_verificacion',
    r'\bspam\b': 'spam',
    r'\bduplicate_of\b': 'duplicado_de',
    r'\bpotential_duplicate_of\b': 'posible_duplicado_de',
    r'\bis_deleted\b': 'esta_eliminado',
    r'\bdeleted_by\b': 'eliminado_por',
    r'\bdeleted_at\b': 'fecha_eliminacion',
    r'\bdeletion_reason\b': 'motivo_eliminacion',
}

def is_safe_to_replace(content):
    # This function processes the whole file, but we only apply generic replacements
    # to files that we KNOW interact with the DB and we explicitly prevent `{ status: 200 }` overrides.
    pass

for root, _, files in os.walk('src'):
    for file in files:
        if file.endswith('.ts') or file.endswith('.tsx'):
            path = os.path.join(root, file)
            with open(path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            original_content = content
            
            # Temporarily protect "status: XXX" or "status:" objects
            protected = {}
            def prot(m):
                k = f"__PROT_{len(protected)}__"
                protected[k] = m.group(0)
                return k
            content = re.sub(r'status:\s*\d+', prot, content)
            content = re.sub(r'status\s*=\s*\d+', prot, content)
            content = re.sub(r'res\.status', prot, content)
            content = re.sub(r'response\.status', prot, content)
            content = re.sub(r'status:\s*\"[^\"]+\"', prot, content) # like status: "success"
            
            # ALSO protect `case.status`, `item.status`, etc if we want frontend to stay working
            content = re.sub(r'item\.status', prot, content)
            content = re.sub(r'\.status\b', prot, content) # actually protect ALL `.status` EXCEPT `row.status`
            
            # Wait, we DO want `row.status` to be replaced because `row` comes from SQL!
            # Let's un-protect `row.status` manually afterwards
            
            for old, new in replacements.items():
                content = re.sub(old, new, content)
                
            # Specifically replace SQL status and source:
            content = re.sub(r'\bstatus\b(?!\s*[:=])', 'estado_actual', content)
            
            # But wait, what if it matches `export type CaseStatus`?
            # It's better to just do manual regex.
            
            pass

